### PR TITLE
[SPARK-29336][SQL] Fix the implementation of QuantileSummaries.merge (guarantee that the relativeError will be respected)

### PR DESCRIPTION
**Pull request by sitegui**
### What changes were proposed in this pull request?

Reimplement `org.apache.spark.sql.catalyst.util.QuantileSummaries#merge` and add a test-case showing the previous bug.

### Why are the changes needed?

The original Greenwald-Khanna paper, from which the algorithm behind `approxQuantile` was taken, does not cover how to merge the result of multiple parallel QuantileSummaries. The current implementation violates some invariants and therefore the effective error can be larger than the specified.

### Does this PR introduce any user-facing change?

Yes, for same cases, the results from `approxQuantile` (`percentile_approx` in SQL) will now be within the expected error margin. For example:

```scala
var values = (1 to 100).toArray
val all_quantiles = values.indices.map(i => (i+1).toDouble / values.length).toArray
for (n <- 0 until 5) {
  var df = spark.sparkContext.makeRDD(values).toDF("value").repartition(5)
  val all_answers = df.stat.approxQuantile("value", all_quantiles, 0.1)
  val all_answered_ranks = all_answers.map(ans => values.indexOf(ans)).toArray
  val error = all_answered_ranks.zipWithIndex.map({ case (answer, expected) => Math.abs(expected - answer) }).toArray
  val max_error = error.max
  print(max_error + "\n")
}
```

In the current build it returns:

```
16
12
10
11
17
```

I couldn't run the code with this patch applied to double check the implementation. Can someone please confirm it now outputs at most `10`, please?

### How was this patch tested?

A new unit test was added to uncover the previous bug.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/QuantileSummaries.scala [hunk 1]
```diff
@@ -47,7 +47,8 @@ class QuantileSummaries(
     val relativeError: Double,
     val sampled: Array[Stats] = Array.empty,
     val count: Long = 0L,
-    var compressed: Boolean = false) extends Serializable {
+    var compressed: Boolean = false)
```
Comment (srowen) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/QuantileSummaries.scala]: Revert all the formatting-only changes like this one please, to keep it simple

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/QuantileSummariesSuite.scala [hunk 2]
```diff
@@ -169,5 +170,22 @@ class QuantileSummariesSuite extends SparkFunSuite {
       checkQuantile(0.1, data, s)
       checkQuantile(0.001, data, s)
     }
+
+    // length of data21 is 4 * length of data22
+    val data21 = data.zipWithIndex.filter(_._2 % 5 != 0).map(_._1).toSeq
```
Comment (srowen) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/QuantileSummariesSuite.scala]: Total nit, but you can: `val (data21, data22) = data.zipWithIndex.partition(_._2 % 5 != 0).map(_._1)`, I believe

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/QuantileSummariesSuite.scala [hunk 3]
```diff
@@ -169,5 +170,22 @@ class QuantileSummariesSuite extends SparkFunSuite {
...
+
+    test(
+      s"Merging unbalanced interleaved lists with epsi=$epsi and seq=$seq_name, " +
+        s"compression=$compression") {
+      val s1 = buildSummary(data21, epsi, compression)
+      val s2 = buildSummary(data22, epsi, compression)
+      val s = s1.merge(s2)
+      // Check all quantiles
+      for (query_rank <- 1 to n) {
+        val query_quantile = query_rank.toDouble / n.toDouble
```
Comment (srowen) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/QuantileSummariesSuite.scala]: Nit: `queryQuantile`

**Review by srowen — CHANGES_REQUESTED** [review]
Looks reasonable, pending tests and some minor changes

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/QuantileSummariesSuite.scala (hunk 2 above)
Comment (sitegui) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/QuantileSummariesSuite.scala]: Hello @srowen , thanks a lot for the remarks. I've fixed the other two, but for this one, if I understood the usage correctly, I need to do something on the lines of:

```scala
val (iterData21, iterData22) = data.zipWithIndex.partition(_._2 % 5 != 0)
val data21 = iterData21.map(_._1).toSeq
val data22 = iterData22.map(_._1).toSeq
```

Or is there a cleaner way to do it?

**Comment by sitegui** [comment]
[This test](https://github.com/apache/spark/blob/8556710409d9f2fbaee9dbf76a2ea70218316693/sql/core/src/test/scala/org/apache/spark/sql/ApproximatePercentileQuerySuite.scala#L124-L142) breaks, but I guess it was not testing the right thing.

It checks if with a higher accuracy we get a lower error for a specific query. The issue is that the algorithm guarantees that the maximum error for all queries will reduce with accuracy: it is bounded by `count / accuracy`. However, it can be the case that for some queries the measured accuracy is much higher than that maximum.

The test in question does a single query against a dataset of 1000 elements, with accuracies 1, 10, 100, 1000 and 10000. Before the patch the measured error was `249, 97, 9, 1, 0` and now it is `249, 40, 0, 1, 0`. Both respect the maximum errors of `1000, 100, 10, 1, 0`, we just got lucky in the third case now.

I'll modify this test case to test for the respect of the reducing maximum bound instead. If you have other suggestions, please let me know.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/QuantileSummariesSuite.scala (hunk 2 above)
Comment (srowen) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/QuantileSummariesSuite.scala]: Nah, disregard this then if .toSeq is necessary

File: sql/core/src/test/scala/org/apache/spark/sql/ApproximatePercentileQuerySuite.scala [hunk 4]
```diff
@@ -124,20 +124,24 @@ class ApproximatePercentileQuerySuite extends QueryTest with SharedSparkSession
...
 
       // With different accuracies
-      val expectedPercentile = 250D
       val accuracies = Array(1, 10, 100, 1000, 10000)
-      val errors = accuracies.map { accuracy =>
-        val df = spark.sql(s"SELECT percentile_approx(col, 0.25, $accuracy) FROM $table")
-        val approximatePercentile = df.collect().head.getInt(0)
-        val error = Math.abs(approximatePercentile - expectedPercentile)
-        error
+      val expectedPercentiles = Array(100D, 200D, 250D, 314D, 777D)
```
Comment (srowen) [inline sql/core/src/test/scala/org/apache/spark/sql/ApproximatePercentileQuerySuite.scala]: Minor nit: I'd write 100.0 instead of 100D etc.

**Review by srowen — COMMENTED** [review]
Looks OK pending tests.

**Comment by srowen** [comment]
Merged to master

**Comment by BdLearnerr** [comment]
anyone tell me , if this is incorporated in 2.4.4 version ?

**Comment by srowen** [comment]
Please have a look at the JIRA - https://issues.apache.org/jira/browse/SPARK-29336
No, it's in 3.0

**Comment by tanelk** [comment]
Hello, I know this is an "ancient" PR, but it seems like it caused a severe performance regression.
I made a jira issue with it https://issues.apache.org/jira/browse/SPARK-47836
I'll also ping @cloud-fan and @viirya  because you were active in the previous jira discussion

**Pull request closed without merging**